presto-spi/src/main/java/com/facebook/presto/spi/type/FloatType.java

```diff
@@ -0,0 +1,121 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.type;
+
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+
+import static com.facebook.presto.spi.StandardErrorCode.NUMERIC_VALUE_OUT_OF_RANGE;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static io.airlift.slice.SizeOf.SIZE_OF_FLOAT;
+
+public final class FloatType
+        extends AbstractFixedWidthType
+{
+    public static final FloatType FLOAT = new FloatType();
+
+    private FloatType()
+    {
+        super(parseTypeSignature(StandardTypes.FLOAT), long.class, SIZE_OF_FLOAT);
+    }
+
+    @Override
+    public boolean isComparable()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isOrderable()
+    {
+        return false;
+    }
+
+    @Override
+    public Object getObjectValue(ConnectorSession session, Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+        return block.getFloat(position, 0);
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        float leftValue = leftBlock.getFloat(leftPosition, 0);
+        float rightValue = rightBlock.getFloat(rightPosition, 0);
+        return leftValue == rightValue;
+    }
+
+    @Override
+    public long hash(Block block, int position)
+    {
+        return (long) Float.floatToRawIntBits(block.getFloat(position, 0));
+    }
+
+    @Override
+    public int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        // WARNING: the correctness of InCodeGenerator is dependent on the implementation of this
+        // function being the equivalence of internal long representation.
+        float leftValue = leftBlock.getFloat(leftPosition, 0);
+        float rightValue = rightBlock.getFloat(rightPosition, 0);
+        return Float.compare(leftValue, rightValue);
+    }
+
+    @Override
+    public void appendTo(Block block, int position, BlockBuilder blockBuilder)
+    {
+        if (block.isNull(position)) {
+            blockBuilder.appendNull();
+        }
+        else {
+            blockBuilder.writeInt(block.getInt(position, 0)).closeEntry();
+        }
+    }
+
+    @Override
+    public long getLong(Block block, int position)
+    {
+        return (long) block.getInt(position, 0);
+    }
+
+    @Override
+    public void writeLong(BlockBuilder blockBuilder, long value)
+    {
+        try {
+            Math.toIntExact(value);
+        }
+        catch (ArithmeticException e) {
+            throw new PrestoException(NUMERIC_VALUE_OUT_OF_RANGE,
```

Comment:
ya, it is more descriptive, but the message will be surfaced to the user. The reason not to use `NUMERIC_VALUE_OUT_OF_RANGE` is that that will be categorized as a "user" error, so anyone auditing the failure logs would miss that that was a bug in Presto. Internally, we tail the query failure log and report anything that fails with an internal error automaticallly